Add support for collecting and sending function logs to APM Server

lahsivjar opened the pull request:
Adds support for collecting `function` logs using AWS Logs API and shipping it to APM-Server with collected agent metadata.

Closes: #256

apmmachine commented:
## :broken_heart: Tests Failed


<!-- BUILD BADGES-->
> _the below badges are clickable and redirect to their specific view in the CI or DOCS_
[![Pipeline View](https://img.shields.io/badge/pipeline-pipeline%20-green)](https://apm-ci.elastic.co/blue/organizations/jenkins/library%2Fapm-aws-lambda-mbp%2FPR-303/detail/PR-303/1//pipeline) [![Test View](https://img.shields.io/badge/test-test-green)](https://apm-ci.elastic.co/blue/organizations/jenkins/library%2Fapm-aws-lambda-mbp%2FPR-303/detail/PR-303/1//tests) [![Changes](https://img.shields.io/badge/changes-changes-green)](https://apm-ci.elastic.co/blue/organizations/jenkins/library%2Fapm-aws-lambda-mbp%2FPR-303/detail/PR-303/1//changes) [![Artifacts](https://img.shields.io/badge/artifacts-artifacts-yellow)](https://apm-ci.elastic.co/blue/organizations/jenkins/library%2Fapm-aws-lambda-mbp%2FPR-303/detail/PR-303/1//artifacts) [![preview](https://img.shields.io/badge/docs-preview-yellowgreen)](http://apm-aws-lambda_303.docs-preview.app.elstc.co/diff)  [![preview](https://img.shields.io/badge/elastic-observability-blue)](https://ci-stats.elastic.co/app/apm/services/apm-ci/transactions/view?rangeFrom=2022-09-13T04:05:37.788Z&rangeTo=2022-09-13T04:25:37.788Z&transactionName=BUILD+library%2Fapm-aws-lambda-mbp%2FPR-%7Bnumber%7D&transactionType=job&latencyAggregationType=avg&traceId=c8ccdf27517ef34b4b740ed79dac89e3&transactionId=2bf0af5ac1c235eb)

<!-- BUILD SUMMARY-->
<details><summary>Expand to view the summary</summary>
<p>

#### Build stats




* Start Time: 2022-09-13T04:15:37.788+0000


* Duration: 7 min 25 sec




#### Test stats :test_tube:

| Test         | Results                         |
| ------------ | :-----------------------------: |
| Failed       | 2  |
| Passed       | 153  |
| Skipped      | 2 |
| Total        | 157   |


</p>
</details>

<!-- TEST RESULTS IF ANY-->

  ### Test errors [![2](https://img.shields.io/badge/2%20-red)](https://apm-ci.elastic.co/blue/organizations/jenkins/library%2Fapm-aws-lambda-mbp%2FPR-303/detail/PR-303/1//tests)
<details><summary>Expand to view the tests failures</summary><p>
    
    
##### `Test / Matrix - PLATFORM = 'macosx && x86_64' / Test / TestLateFlush – github.com/elastic/apm-aws-lambda`


<ul>
<details><summary>Expand to view the error details</summary><p>

```
 Failed 
 ```
</p></details>
<details><summary>Expand to view the stacktrace</summary><p>

```
 === RUN   TestLateFlush
{"level":"error","ts":1663042903.545444,"caller":"apm-aws-lambda/main_test.go:381","msg":"Could not send log event : Post \"http://127.0.0.1:49248\": http: server closed idle connection","ecs.version":"1.6.0","stacktrace":"github.com/elastic/apm-aws-lambda.sendLogEvent\n\t/Users/admin/workspace/ibrary_apm-aws-lambda-mbp_PR-303/src/github.com/elastic/apm-aws-lambda/main_test.go:381\ngithub.com/elastic/apm-aws-lambda.processMockEvent\n\t/Users/admin/workspace/ibrary_apm-aws-lambda-mbp_PR-303/src/github.com/elastic/apm-aws-lambda/main_test.go:219"}
{"level":"error","ts":1663042903.545533,"caller":"apm-aws-lambda/main_test.go:381","msg":"Could not send log event : Post \"http://127.0.0.1:49248\": dial tcp 127.0.0.1:49248: connect: connection refused","ecs.version":"1.6.0","stacktrace":"github.com/elastic/apm-aws-lambda.sendLogEvent\n\t/Users/admin/workspace/ibrary_apm-aws-lambda-mbp_PR-303/src/github.com/elastic/apm-aws-lambda/main_test.go:381\ngithub.com/elastic/apm-aws-lambda.processMockEvent\n\t/Users/admin/workspace/ibrary_apm-aws-lambda-mbp_PR-303/src/github.com/elastic/apm-aws-lambda/main_test.go:314"}
{"level":"error","ts":1663042903.5456908,"caller":"apm-aws-lambda/main_test.go:381","msg":"Could not send log event : Post \"http://127.0.0.1:49248\": dial tcp 127.0.0.1:49248: connect: connection refused","ecs.version":"1.6.0","stacktrace":"github.com/elastic/apm-aws-lambda.sendLogEvent\n\t/Users/admin/workspace/ibrary_apm-aws-lambda-mbp_PR-303/src/github.com/elastic/apm-aws-lambda/main_test.go:381\ngithub.com/elastic/apm-aws-lambda.processMockEvent\n\t/Users/admin/workspace/ibrary_apm-aws-lambda-mbp_PR-303/src/github.com/elastic/apm-aws-lambda/main_test.go:317"}
{"log.level":"debug","@timestamp":"2022-09-12T21:21:43.546-0700","log.origin":{"file.name":"extension/client.go","file.line":123},"message":"ExtensionID : b03a29ec-ee63-44cd-8e53-3987a8e8aa8e","ecs.version":"1.6.0"}
{"log.level":"debug","@timestamp":"2022-09-12T21:21:43.546-0700","log.origin":{"file.name":"app/run.go","file.line":47},"message":"Register response: {\n\t\"functionName\": \"UnitTestingMockLambda\",\n\t\"functionVersion\": \"$LATEST\",\n\t\"handler\": \"main_test.mock_lambda\"\n}","ecs.version":"1.6.0"}
{"log.level":"info","@timestamp":"2022-09-12T21:21:43.546-0700","log.origin":{"file.name":"apmproxy/receiver.go","file.line":57},"message":"Extension listening for apm data on :49267","ecs.version":"1.6.0"}
{"log.level":"info","@timestamp":"2022-09-12T21:21:43.546-0700","log.origin":{"file.name":"logsapi/subscribe.go","file.line":73},"message":"Extension listening for Lambda Logs API events on 127.0.0.1:49263","ecs.version":"1.6.0"}
{"log.level":"info","@timestamp":"2022-09-12T21:21:43.546-0700","log.origin":{"file.name":"app/run.go","file.line":138},"message":"Waiting for next event...","ecs.version":"1.6.0"}
{"log.level":"debug","@timestamp":"2022-09-12T21:21:43.546-0700","log.origin":{"file.name":"app/run.go","file.line":155},"message":"Received event.","ecs.version":"1.6.0"}
{"log.level":"debug","@timestamp":"2022-09-12T21:21:43.546-0700","log.origin":{"file.name":"app/run.go","file.line":156},"message":"{\n\t\"timestamp\": \"2022-09-12T21:21:43.54689-07:00\",\n\t\"eventType\": \"INVOKE\",\n\t\"deadlineMs\": 1663042908546,\n\t\"requestId\": \"3b79cf94-7793-4612-a653-34452d5236c6\",\n\t\"invokedFunctionArn\": \"arn:aws:lambda:eu-central-1:627286350134:function:main_unit_test\",\n\t\"tracing\": {\n\t\t\"type\": \"\",\n\t\t\"value\": \"\"\n\t}\n}","ecs.version":"1.6.0"}
{"log.level":"debug","@timestamp":"2022-09-12T21:21:43.548-0700","log.origin":{"file.name":"logsapi/event.go","file.line":84},"message":"Received log event platform.start","ecs.version":"1.6.0"}
{"log.level":"debug","@timestamp":"2022-09-12T21:21:43.548-0700","log.origin":{"file.name":"logsapi/event.go","file.line":84},"message":"Received log event platform.runtimeDone","ecs.version":"1.6.0"}
{"log.level":"info","@timestamp":"2022-09-12T21:21:43.548-0700","log.origin":{"file.name":"logsapi/event.go","file.line":90},"message":"Received runtimeDone event for this function invocation","ecs.version":"1.6.0"}
{"log.level":"debug","@timestamp":"2022-09-12T21:21:43.548-0700","log.origin":{"file.name":"app/run.go","file.line":206},"message":"Received runtimeDone signal","ecs.version":"1.6.0"}
{"log.level":"debug","@timestamp":"2022-09-12T21:21:43.548-0700","log.origin":{"file.name":"app/run.go","file.line":112},"message":"Waiting for background data send to end","ecs.version":"1.6.0"}
{"log.level":"debug","@timestamp":"2022-09-12T21:21:43.548-0700","log.origin":{"file.name":"apmproxy/apmserver.go","file.line":53},"message":"Invocation context cancelled, not processing any more agent data","ecs.version":"1.6.0"}
{"log.level":"debug","@timestamp":"2022-09-12T21:21:43.548-0700","log.origin":{"file.name":"apmproxy/apmserver.go","file.line":76},"message":"Flush started - Checking for agent data","ecs.version":"1.6.0"}
{"log.level":"debug","@timestamp":"2022-09-12T21:21:43.548-0700","log.origin":{"file.name":"apmproxy/apmserver.go","file.line":85},"message":"Flush ended - No agent data on buffer","ecs.version":"1.6.0"}
{"log.level":"info","@timestamp":"2022-09-12T21:21:43.548-0700","log.origin":{"file.name":"app/run.go","file.line":138},"message":"Waiting for next event...","ecs.version":"1.6.0"}
{"log.level":"debug","@timestamp":"2022-09-12T21:21:43.548-0700","log.origin":{"file.name":"apmproxy/receiver.go","file.line":115},"message":"Handling APM Data Intake","ecs.version":"1.6.0"}
{"log.level":"debug","@timestamp":"2022-09-12T21:21:43.548-0700","log.origin":{"file.name":"apmproxy/apmserver.go","file.line":289},"message":"Adding agent data to buffer to be sent to apm server","ecs.version":"1.6.0"}
{"log.level":"debug","@timestamp":"2022-09-12T21:21:43.549-0700","log.origin":{"file.name":"app/run.go","file.line":155},"message":"Received event.","ecs.version":"1.6.0"}
{"log.level":"debug","@timestamp":"2022-09-12T21:21:43.549-0700","log.origin":{"file.name":"app/run.go","file.line":156},"message":"{\n\t\"timestamp\": \"2022-09-12T21:21:43.549092-07:00\",\n\t\"eventType\": \"INVOKE\",\n\t\"deadlineMs\": 1663042908549,\n\t\"requestId\": \"f0906330-bfa7-4037-973f-fb4ac93259b6\",\n\t\"invokedFunctionArn\": \"arn:aws:lambda:eu-central-1:627286350134:function:main_unit_test\",\n\t\"tracing\": {\n\t\t\"type\": \"\",\n\t\t\"value\": \"\"\n\t}\n}","ecs.version":"1.6.0"}
{"log.level":"debug","@timestamp":"2022-09-12T21:21:43.549-0700","log.origin":{"file.name":"app/run.go","file.line":204},"message":"APM client has pending flush signals","ecs.version":"1.6.0"}
{"log.level":"debug","@timestamp":"2022-09-12T21:21:43.549-0700","log.origin":{"file.name":"app/run.go","file.line":112},"message":"Waiting for background data send to end","ecs.version":"1.6.0"}
{"log.level":"debug","@timestamp":"2022-09-12T21:21:43.549-0700","log.origin":{"file.name":"logsapi/event.go","file.line":120},"message":"Current invocation over. Interrupting logs processing goroutine","ecs.version":"1.6.0"}
{"log.level":"debug","@timestamp":"2022-09-12T21:21:43.550-0700","log.origin":{"file.name":"apmproxy/apmserver.go","file.line":141},"message":"Sending data chunk to APM server","ecs.version":"1.6.0"}
{"level":"debug","ts":1663042903.551054,"caller":"apm-aws-lambda/main_test.go:106","msg":"Event type received by mock APM server : TimelyResponse","ecs.version":"1.6.0"}
{"level":"debug","ts":1663042903.5510652,"caller":"apm-aws-lambda/main_test.go:109","msg":"Timely response signal received","ecs.version":"1.6.0"}
{"level":"debug","ts":1663042903.551069,"caller":"apm-aws-lambda/main_test.go:128","msg":"APM Payload processed","ecs.version":"1.6.0"}
{"log.level":"debug","@timestamp":"2022-09-12T21:21:43.551-0700","log.origin":{"file.name":"apmproxy/apmserver.go","file.line":234},"message":"APM server Transport status set to Healthy","ecs.version":"1.6.0"}
{"log.level":"debug","@timestamp":"2022-09-12T21:21:43.551-0700","log.origin":{"file.name":"apmproxy/apmserver.go","file.line":53},"message":"Invocation context cancelled, not processing any more agent data","ecs.version":"1.6.0"}
{"log.level":"debug","@timestamp":"2022-09-12T21:21:43.551-0700","log.origin":{"file.name":"apmproxy/apmserver.go","file.line":76},"message":"Flush started - Checking for agent data","ecs.version":"1.6.0"}
{"log.level":"debug","@timestamp":"2022-09-12T21:21:43.551-0700","log.origin":{"file.name":"apmproxy/apmserver.go","file.line":85},"message":"Flush ended - No agent data on buffer","ecs.version":"1.6.0"}
{"log.level":"info","@timestamp":"2022-09-12T21:21:43.551-0700","log.origin":{"file.name":"app/run.go","file.line":138},"message":"Waiting for next event...","ecs.version":"1.6.0"}
{"log.level":"debug","@timestamp":"2022-09-12T21:21:43.551-0700","log.origin":{"file.name":"app/run.go","file.line":155},"message":"Received event.","ecs.version":"1.6.0"}
{"log.level":"debug","@timestamp":"2022-09-12T21:21:43.551-0700","log.origin":{"file.name":"app/run.go","file.line":156},"message":"{\n\t\"timestamp\": \"2022-09-12T21:21:43.551679-07:00\",\n\t\"eventType\": \"SHUTDOWN\",\n\t\"deadlineMs\": 1663042903551,\n\t\"requestId\": \"e95a2d3d-ac34-4c41-8e92-64a0453ca160\",\n\t\"invokedFunctionArn\": \"arn:aws:lambda:eu-central-1:627286350134:function:main_unit_test\",\n\t\"tracing\": {\n\t\t\"type\": \"\",\n\t\t\"value\": \"\"\n\t}\n}","ecs.version":"1.6.0"}
{"log.level":"info","@timestamp":"2022-09-12T21:21:43.551-0700","log.origin":{"file.name":"app/run.go","file.line":109},"message":"Received shutdown event: . Exiting...","ecs.version":"1.6.0"}
{"log.level":"debug","@timestamp":"2022-09-12T21:21:43.551-0700","log.origin":{"file.name":"apmproxy/apmserver.go","file.line":76},"message":"Flush started - Checking for agent data","ecs.version":"1.6.0"}
{"log.level":"debug","@timestamp":"2022-09-12T21:21:43.551-0700","log.origin":{"file.name":"apmproxy/apmserver.go","file.line":85},"message":"Flush ended - No agent data on buffer","ecs.version":"1.6.0"}
{"log.level":"debug","@timestamp":"2022-09-12T21:21:43.551-0700","log.origin":{"file.name":"apmproxy/receiver.go","file.line":61},"message":"server closed","ecs.version":"1.6.0"}
    main_test.go:454: 
        	Error Trace:	main_test.go:454
        	Error:      	"TimelyResponse" does not contain "TimelyResponseTimelyResponse"
        	Test:       	TestLateFlush
--- FAIL: TestLateFlush (0.01s)
 
 ```
</p></details>
</ul>
    
##### `Test / Matrix - PLATFORM = 'macosx && x86_64' / Test / TestWaitGroup – github.com/elastic/apm-aws-lambda`


<ul>
<details><summary>Expand to view the error details</summary><p>

```
 Failed 
 ```
</p></details>
<details><summary>Expand to view the stacktrace</summary><p>

```
 === RUN   TestWaitGroup
{"level":"error","ts":1663042903.552286,"caller":"apm-aws-lambda/main_test.go:381","msg":"Could not send log event : Post \"http://127.0.0.1:49263\": read tcp 127.0.0.1:49269->127.0.0.1:49263: read: connection reset by peer","ecs.version":"1.6.0","stacktrace":"github.com/elastic/apm-aws-lambda.sendLogEvent\n\t/Users/admin/workspace/ibrary_apm-aws-lambda-mbp_PR-303/src/github.com/elastic/apm-aws-lambda/main_test.go:381\ngithub.com/elastic/apm-aws-lambda.processMockEvent\n\t/Users/admin/workspace/ibrary_apm-aws-lambda-mbp_PR-303/src/github.com/elastic/apm-aws-lambda/main_test.go:219"}
{"level":"error","ts":1663042903.552404,"caller":"apm-aws-lambda/main_test.go:381","msg":"Could not send log event : Post \"http://127.0.0.1:49263\": dial tcp 127.0.0.1:49263: connect: connection refused","ecs.version":"1.6.0","stacktrace":"github.com/elastic/apm-aws-lambda.sendLogEvent\n\t/Users/admin/workspace/ibrary_apm-aws-lambda-mbp_PR-303/src/github.com/elastic/apm-aws-lambda/main_test.go:381\ngithub.com/elastic/apm-aws-lambda.processMockEvent\n\t/Users/admin/workspace/ibrary_apm-aws-lambda-mbp_PR-303/src/github.com/elastic/apm-aws-lambda/main_test.go:314"}
{"level":"error","ts":1663042903.552623,"caller":"apm-aws-lambda/main_test.go:381","msg":"Could not send log event : Post \"http://127.0.0.1:49263\": dial tcp 127.0.0.1:49263: connect: connection refused","ecs.version":"1.6.0","stacktrace":"github.com/elastic/apm-aws-lambda.sendLogEvent\n\t/Users/admin/workspace/ibrary_apm-aws-lambda-mbp_PR-303/src/github.com/elastic/apm-aws-lambda/main_test.go:381\ngithub.com/elastic/apm-aws-lambda.processMockEvent\n\t/Users/admin/workspace/ibrary_apm-aws-lambda-mbp_PR-303/src/github.com/elastic/apm-aws-lambda/main_test.go:317"}
{"log.level":"debug","@timestamp":"2022-09-12T21:21:43.553-0700","log.origin":{"file.name":"extension/client.go","file.line":123},"message":"ExtensionID : b03a29ec-ee63-44cd-8e53-3987a8e8aa8e","ecs.version":"1.6.0"}
{"log.level":"debug","@timestamp":"2022-09-12T21:21:43.553-0700","log.origin":{"file.name":"app/run.go","file.line":47},"message":"Register response: {\n\t\"functionName\": \"UnitTestingMockLambda\",\n\t\"functionVersion\": \"$LATEST\",\n\t\"handler\": \"main_test.mock_lambda\"\n}","ecs.version":"1.6.0"}
{"log.level":"info","@timestamp":"2022-09-12T21:21:43.553-0700","log.origin":{"file.name":"apmproxy/receiver.go","file.line":57},"message":"Extension listening for apm data on :49280","ecs.version":"1.6.0"}
{"log.level":"info","@timestamp":"2022-09-12T21:21:43.553-0700","log.origin":{"file.name":"logsapi/subscribe.go","file.line":73},"message":"Extension listening for Lambda Logs API events on 127.0.0.1:49276","ecs.version":"1.6.0"}
{"log.level":"info","@timestamp":"2022-09-12T21:21:43.553-0700","log.origin":{"file.name":"app/run.go","file.line":138},"message":"Waiting for next event...","ecs.version":"1.6.0"}
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x10 pc=0x13760ae]

goroutine 166 [running]:
github.com/elastic/apm-aws-lambda.processMockEvent({0xc000032750, 0x24}, {{0x14345e0, 0x8}, {0x1436ba0, 0xe}, 0x0, 0x4014000000000000}, {0xc00013c3e6, 0x5}, ...)
	/Users/admin/workspace/ibrary_apm-aws-lambda-mbp_PR-303/src/github.com/elastic/apm-aws-lambda/main_test.go:246 +0x34e
created by github.com/elastic/apm-aws-lambda.newMockLambdaServer.func1
	/Users/admin/workspace/ibrary_apm-aws-lambda-mbp_PR-303/src/github.com/elastic/apm-aws-lambda/main_test.go:178 +0x415
 
 ```
</p></details>
</ul>
    
</p></details>


<!-- STEPS ERRORS IF ANY -->














### Steps errors [![1](https://img.shields.io/badge/1%20-red)](https://apm-ci.elastic.co/blue/organizations/jenkins/library%2Fapm-aws-lambda-mbp%2FPR-303/detail/PR-303/1//pipeline)
  <details><summary>Expand to view the steps failures</summary>
  <p>


  
  
  
  
##### `Running Go tests`
<ul>
<li>Took 0 min 24 sec . View more details <a href="https://apm-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/library/pipelines/apm-aws-lambda-mbp/pipelines/PR-303/runs/1/steps/255/log/?start=0">here</a></li>
<li>Description: <code>gotestsum --format testname --junitfile junit-report.xml -- -v ./...</code></l1>
</ul>
  
  </p>
  </details>



## :robot: GitHub comments

To re-run your PR in the CI, just comment with:



- `/test` : Re-trigger the build.


- `run` `elasticsearch-ci/docs` : Re-trigger the docs validation. (use unformatted text in the comment!)




<!--COMMENT_GENERATED_WITH_ID_comment.id-->

lahsivjar commented:
/test

felixbarny commented on logsapi/functionlogs.go:
```diff
@@ -0,0 +1,99 @@
+	"go.elastic.co/fastjson"
+)
+
+type LogContainer struct {
+	Log *logLine
+}
+
+type logLine struct {
+	Timestamp model.Time
+	Message   string
+	FAAS      *model.FAAS
+}
+
+func (l *logLine) MarshalFastJSON(w *fastjson.Writer) error {
```
Just an idea that could reduce allocations: instead of converting `extension.NextEventResponse` and `LogEvent` into a `LogContainer`, you could pass in `extension.NextEventResponse` and `LogEvent` directly to the `MarshalFastJSON` method and avoid creating these temporary serialization objects.

However, if you feel like the code is cleaner this way and if this isn't contributing a lot to overall allocations, feel free to ignore.

lahsivjar replied:
I ran some benchmarks on the idea; below are the results:

```
# With `LogContainer` and `logLine`
BenchmarkProcessFunctionLog-10    	 1758667	       684.3 ns/op	    1264 B/op	       7 allocs/op

# Without `LogContainer` and `logLine`
BenchmarkProcessFunctionLog-10    	 1808710	       668.9 ns/op	    1184 B/op	       6 allocs/op
```

Based on the results I am a bit more inclined toward keeping the objects since they will allow for easier extension in case more fields need to be added.

WDYT given the benchmark results?

<details>
<summary>(Code diff that produced the 2nd benchmark)</summary>

```diff
diff --cc logsapi/functionlogs.go
index 984b15f,3213428..0000000
--- a/logsapi/functionlogs.go
+++ b/logsapi/functionlogs.go
@@@ -24,28 -24,61 +24,23 @@@ import 
  	"go.elastic.co/fastjson"
  )
  
- type logContainer struct {
 -type LogContainer struct {
--	Log *logLine
--}
--
--type logLine struct {
--	Timestamp model.Time
--	Message   string
--	FAAS      *model.FAAS
--}
--
--func (l *logLine) MarshalFastJSON(w *fastjson.Writer) error {
 -	var firstErr error
 -	w.RawByte('{')
 -	w.RawString("\"message\":")
 -	w.String(l.Message)
 -	w.RawString(",\"@timestamp\":")
 -	if err := l.Timestamp.MarshalFastJSON(w); err != nil && firstErr == nil {
 -		firstErr = err
 -	}
 -	if l.FAAS != nil {
 -		w.RawString(",\"faas\":")
 -		if err := l.FAAS.MarshalFastJSON(w); err != nil && firstErr == nil {
 -			firstErr = err
 -		}
 -	}
 -	w.RawByte('}')
 -	return firstErr
 -}
 -
 -func (lc LogContainer) MarshalFastJSON(json *fastjson.Writer) error {
 -	json.RawString(`{"log":`)
 -	if err := lc.Log.MarshalFastJSON(json); err != nil {
 -		return err
 -	}
 -	json.RawString(`}`)
 -	return nil
 -}
 -
+ func marshalFastJSON(w *fastjson.Writer, fd *extension.NextEventResponse, log LogEvent) error {
 -	var firstErr error // 4
 +	var firstErr error
+ 	w.RawString(`{"log":`)
  	w.RawByte('{')
  	w.RawString("\"message\":")
- 	w.String(l.Message)
+ 	w.String(log.StringRecord)
  	w.RawString(",\"@timestamp\":")
- 	if err := l.Timestamp.MarshalFastJSON(w); err != nil && firstErr == nil {
+ 	if err := model.Time(log.Time).MarshalFastJSON(w); err != nil && firstErr == nil {
  		firstErr = err
  	}
- 	if l.FAAS != nil {
+ 	if fd != nil {
  		w.RawString(",\"faas\":")
- 		if err := l.FAAS.MarshalFastJSON(w); err != nil && firstErr == nil {
+ 		faas := model.FAAS{
+ 			Execution: fd.RequestID,
+ 			ID:        fd.InvokedFunctionArn,
+ 		}
+ 		if err := faas.MarshalFastJSON(w); err != nil && firstErr == nil {
  			firstErr = err
  		}
  	}
@@@ -69,23 -94,8 +56,8 @@@ func ProcessFunctionLog
  	functionData *extension.NextEventResponse,
  	log LogEvent,
  ) (apmproxy.AgentData, error) {
- 	lc := logContainer{
- 		Log: &logLine{
- 			Timestamp: model.Time(log.Time),
- 			Message:   log.StringRecord,
- 		},
- 	}
- 
- 	if functionData != nil {
- 		// FaaS Fields
- 		lc.Log.FAAS = &model.FAAS{
- 			Execution: functionData.RequestID,
- 			ID:        functionData.InvokedFunctionArn,
- 		}
- 	}
- 
 -	var jsonWriter fastjson.Writer // 1
 +	var jsonWriter fastjson.Writer
- 	if err := lc.MarshalFastJSON(&jsonWriter); err != nil {
+ 	if err := marshalFastJSON(&jsonWriter, functionData, log); err != nil {
  		return apmproxy.AgentData{}, err
  	}
  
diff --git a/logsapi/functionlogs_test.go b/logsapi/functionlogs_test.go
index 64d4312..5aed6a6 100644
--- a/logsapi/functionlogs_test.go
+++ b/logsapi/functionlogs_test.go
@@ -68,3 +68,23 @@ func TestProcessFunctionLog(t *testing.T) {
 		)
 	})
 }
+
+func BenchmarkProcessFunctionLog(b *testing.B) {
+	mc := &apmproxy.MetadataContainer{
+		Metadata: []byte(`{"metadata":{"service":{"agent":{"name":"apm-lambda-extension","version":"1.1.0"},"framework":{"name":"AWS Lambda","version":""},"language":{"name":"python","version":"3.9.8"},"runtime":{"name":"","version":""},"node":{}},"user":{},"process":{"pid":0},"system":{"container":{"id":""},"kubernetes":{"node":{},"pod":{}}},"cloud":{"provider":"","instance":{},"machine":{},"account":{},"project":{},"service":{}}}}`),
+	}
+	event := LogEvent{
+		Time:         time.Date(2022, 11, 12, 0, 0, 0, 0, time.UTC),
+		Type:         FunctionLog,
+		StringRecord: "ERROR encountered. Stack trace:\nmy-function (line 10)\n",
+	}
+	nextEventResp := &extension.NextEventResponse{
+		RequestID:          "8476a536-e9f4-11e8-9739-2dfe598c3fcd",
+		InvokedFunctionArn: "arn:aws:lambda:us-east-2:123456789012:function:custom-runtime",
+	}
+
+	for n := 0; n < b.N; n++ {
+		_, err := ProcessFunctionLog(mc, nextEventResp, event)
+		require.NoError(b, err)
+	}
+}
```
</details>

felixbarny replied:
Seems to not make a huge difference allocation-wise.

> Based on the results I am a bit more inclined toward keeping the objects since they will allow for easier extension in case more fields need to be added.

SGTM

lahsivjar commented on logsapi/event.go:
```diff
@@ -19,38 +19,44 @@ package logsapi
 )
 
-// EventType represents the type of logs in Lambda
-type EventType string
```
[**For reviewers**]: I have renamed this from `EventType` -> `LogType` to avoid confusion with the `EventType` used for lifecycle events. Let me know if it doesn't sound correct.

lahsivjar commented on logsapi/event.go:
```diff
@@ -101,6 +108,18 @@ func (lc *Client) ProcessLogs(
 					lc.logger.Debug("Log API runtimeDone event request id didn't match")
 				}
+			case FunctionLog:
+				if metadataContainer != nil && len(metadataContainer.Metadata) > 0 {
```
[**For reviewers**]: The way we collect metadata currently allows function logs to be emitted before metadata can be collected. Since APM Server considers any request without metadata to be `400` the PR drops any log lines before metadata is collected. Another approach can be to buffer these log lines and send them in batch but that might increase the resource utilization of the extension. Let me know if there are any concerns/alternative approaches here.

lahsivjar commented on logsapi/functionlogs_test.go:
```diff
@@ -0,0 +1,54 @@
+	reqID := "8476a536-e9f4-11e8-9739-2dfe598c3fcd"
+	invokedFnArn := "arn:aws:lambda:us-east-2:123456789012:function:custom-runtime"
+	expectedData := fmt.Sprintf(
+		"%s\n{\"log\":{\"message\":\"%s\",\"@timestamp\":%d,\"faas\":{\"coldstart\":false,\"execution\":\"%s\",\"id\":\"%s\"}}}",
```
[**For reviewers**] `coldstart` value here might not be correct and (CMIIW) I don't think there is a way for us to know about this by using `function` log event type. However, the default marshal logic always adds `coldstart` field.

I am thinking it might be better for us to have a custom unmarshal for `FAAS` here. Happy to discuss other thoughts on this.

lahsivjar replied:
While testing the feature, I noticed that it is possible to get `function` log events while a different request ID is being processed. The payload of `function` log doesn't include any `requestID` for correlation with other events.

Based on current knowledge, I think we will have to drop `FAAS` fields from the log body completely.

lahsivjar replied:
For posterity, discussed this with @axw and concluded with the following:

1. Add a custom Marshal logic for `faas`
2. Use `platform.start` and `platform.end` events to track request ID for function logs (assuming function logs will always be bounded between these two events)

felixbarny replied:
that sounds like a smart solution!

lahsivjar replied:
This is fixed now.

axw commented on logsapi/event.go:
```diff
@@ -101,6 +99,23 @@ func (lc *Client) ProcessLogs(
 					lc.logger.Debug("Log API runtimeDone event request id didn't match")
 				}
+			case FunctionLog:
+				if metadataContainer != nil && len(metadataContainer.Metadata) > 0 {
```
Could we neaten this a little by doing this?

```go
if metadataContainer == nil || len(metadataContainer.Metadata) == 0 {
    lc.logger.Warn("Function log received before metadata populated, quietly dropping log")
    continue
}

lc.logger.Debug("Received function log")
...
```

lahsivjar replied:
I have moved the nil handling inside `ProcessFunctionLog`, felt it would be a bit cleaner this way. Let me know your thoughts.

axw replied:
Looks good, thanks!

axw commented on logsapi/functionlogs.go:
```diff
@@ -0,0 +1,118 @@
+//
+// The purpose of having a separate struct is to have a custom
+// marshalling logic that is targeted for the faas fields
+// available for function logs. For example: `coldstart` value
+// cannot be inferred for function logs so this struct drops
+// the field entirely.
```
Doesn't the extension get initialised once per cold start? Meaning that we could set `coldstart: true` for the first invocation processed by the extension, and `coldstart: false` for every request thereafter?

We don't need to change this now, but I think we _could_ improve it in the future. We probably don't need to set this property on every log message anyway, though.

lahsivjar replied:
> Doesn't the extension get initialised once per cold start? Meaning that we could set coldstart: true for the first invocation processed by the extension, and coldstart: false for every request thereafter?

Hmm, that's an amazing idea. I will create an issue for this.

lahsivjar commented on main.go:
```diff
@@ -39,12 +40,19 @@ func main() {
-	)
+	}
+
+	disableFnLog, _ := strconv.ParseBool(os.Getenv("ELASTIC_DISABLE_FUNCTION_LOG_SUBSCRIPTION"))
```
Initially, I kept a negative flag to keep it `false` by default since not providing the config seemed more aligned with a `false` value. Agree with double-negs being confusing to users, I have updated it to use `ELASTIC_APM_LAMBDA_CAPTURE_LOGS`.

lahsivjar commented:
@axw Apologies for the force push 🙇 

[This is the commit](https://github.com/elastic/apm-aws-lambda/pull/303/commits/7e6865cb61e2b52e29f09cc2944ee186bfeaa7e0) added for addressing the review comments.

axw reviewed (approved):
Looks great, thanks for the updates :)

lahsivjar commented on logsapi/functionlogs.go:
```diff
@@ -0,0 +1,118 @@
+}
+
+// ProcessFunctionLog consumes extension event, agent metadata and log
+// event from Lambda logs API to create a payload for APM server
```
Created an issue: https://github.com/elastic/apm-aws-lambda/issues/311

kruskall commented:
Tested with `BC2` (and `8.6.0-SNAPSHOT`) by provisioning the environment with the terraform script (`SKIP_DESTROY=1` has been used to inspect the cluster) with `TF_VAR_ess_version=8.5.0-SNAPSHOT`.

Verified that the logs are visible in the logs tab in the Observability section.

Made sure that `ELASTIC_APM_LAMBDA_CAPTURE_LOGS=false` disables logs collection.

Possible concern: https://github.com/elastic/apm-aws-lambda/issues/322
I don't think the above issue is a blocker according to the definition of a blocker so I'm going to label this PR as test-plan-ok

RyanHealey1307 commented:
I'm aware this is still experimental, but some feedback...
I've tested this for a python lambda, using the latest APM layer, and when you invocate the function, the logs don't go through to APM until the next invocation. Eg, if my lambda is run at 10:09 and then again at 10:13, the logs from 10:09 are not visible until next next invocation at 10:13. My timeout is set to 5 mins and it's a very simple function, so I can't be a timeout issue.
